### PR TITLE
*: reduce analyze table read when table contains PK

### PR DESCRIPTION
**Pull request by lamxTyler**
This PR reduce the analyze table read times from 2 to 1 when table contains PK by build histograms directly for PK when sampling. In order to do so, the building process for PK and index is changed to a more flexible way by separating the process into initialization and incremental construction.

PTAL @shenli @hanfei1991 @winoros

### CODE REVIEW

File: statistics/builder.go [hunk 1]
```diff
@@ -17,29 +17,104 @@ import (
...
+	}
 }
 
-func build4SortedColumn(ctx context.Context, numBuckets, id int64, records ast.RecordSet, isPK bool) (int64, *Histogram, error) {
-	hg := &Histogram{
-		ID:      id,
-		NDV:     0,
-		Buckets: make([]Bucket, 1, numBuckets),
+// Insert inserts one row into the histogram.
+func (b *SortedBuilder) Insert(row *ast.Row) error {
```
Comment (hanfei1991) [inline statistics/builder.go]: Iterate is a better name.

File: executor/analyze.go [hunk 2]
```diff
@@ -163,17 +161,18 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultCh chan<-
...
 func (e *AnalyzeExec) analyzeColumns(task *analyzeTask) analyzeResult {
-	collectors, err := CollectSamplesAndEstimateNDVs(&recordSet{executor: task.src}, len(task.Columns))
+	collectors, pkBuilder, err := CollectSamplesAndEstimateNDVs(e.ctx, &recordSet{executor: task.src}, len(task.Columns), task.PKInfo)
 	if err != nil {
 		return analyzeResult{err: err}
 	}
-	result := analyzeResult{tableID: task.tableInfo.ID, count: collectors[0].Count + collectors[0].NullCount, isIndex: 0}
+	result := analyzeResult{tableID: task.tableInfo.ID, isIndex: 0}
+	if task.PKInfo != nil {
+		result.count = int64(pkBuilder.Hist.TotalRowCount())
```
Comment (hanfei1991) [inline executor/analyze.go]: Why not return count directly ?

File: executor/analyze.go [hunk 3]
```diff
@@ -163,17 +161,18 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultCh chan<-
...
+		result.hist = append(result.hist, pkBuilder.Hist)
```
Comment (hanfei1991) [inline executor/analyze.go]: It's simpler to assign []*Histogram{pkBuilder.Hist} directly, isn't it ?

File: executor/analyze.go [hunk 4]
```diff
@@ -230,15 +234,22 @@ func CollectSamplesAndEstimateNDVs(e ast.RecordSet, numCols int) ([]*SampleColle
...
 		if err != nil {
-			return nil, errors.Trace(err)
+			return nil, nil, errors.Trace(err)
 		}
 		if row == nil {
-			return collectors, nil
+			return collectors, pkBuilder, nil
+		}
+		if pkInfo != nil {
+			err := pkBuilder.Insert(&ast.Row{Data: []types.Datum{row.Data[0]}})
```
Comment (hanfei1991) [inline executor/analyze.go]: Can we try to pass a slice of datums instead of create a new structure ?

Comment (lamxTyler) [inline executor/analyze.go]: Yes, that seems more reasonable.

File: statistics/builder.go [hunk 5]
```diff
@@ -17,29 +17,104 @@ import (
...
+		},
+	}
 }
 
-func build4SortedColumn(ctx context.Context, numBuckets, id int64, records ast.RecordSet, isPK bool) (int64, *Histogram, error) {
-	hg := &Histogram{
-		ID:      id,
-		NDV:     0,
-		Buckets: make([]Bucket, 1, numBuckets),
+// Iterate inserts row into the histogram one by one.
```
Comment (hanfei1991) [inline statistics/builder.go]: `inserts` may not be accurate. `updates ... incremently` may be better

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**